File: leetcode/medium/307_NumArray.ts
class SegmentTree {
    nums: number[];
    sums: number[];

    constructor (nums: number[]) {
        this.nums = nums;

        const height: number = this.getArrHeight(this.nums.length);
        this.sums = new Array(height).fill(0);

        this.constructTree(0, this.nums.length - 1, 0);
    }

    private getArrHeight (length: number) {
        const height: number = Math.ceil(Math.log2(length));
        return 2 * (2**height) - 1;
    }
    private getLeftChild (id: number) {
        return id * 2 + 1;
    }
    private getRightChild (id: number) {
        return id * 2 + 2;
    }
    private getMiddle (s: number, e: number) {
        return Math.floor((e + s) / 2);
    }
    private constructTree (start: number, end: number, sumId: number) {
        if (start === end) {
            this.sums[sumId] = this.nums[start];
            return this.sums[sumId];
        }

        const middle: number = this.getMiddle(start, end);
        const leftSum: number = this.constructTree(start, middle, this.getLeftChild(sumId));
        const rightSum: number = this.constructTree(middle + 1, end, this.getRightChild(sumId));

        this.sums[sumId] = leftSum + rightSum;
        return this.sums[sumId];
    }
    private update (start: number, end: number, sumId: number, id: number, diff: number) {
        if (id < start || id > end) {
            return;
        }

        this.sums[sumId] += diff;

        if (start === end) {
            return;
        }

        const middle: number = this.getMiddle(start, end);
        const leftChild: number = this.getLeftChild(sumId);
        const rightChild: number = this.getRightChild(sumId);

        this.update(start, middle, leftChild, id, diff);
        this.update(middle + 1, end, rightChild, id, diff);
    }
    private getSum (start: number, end: number, sumId: number, queryStart: number, queryEnd: number): number {
        if (start >= queryStart && end <= queryEnd) {
            return this.sums[sumId];
        }

        if (start > queryEnd || end < queryStart) {
            return 0;
        }

        const middle: number = this.getMiddle(start, end);
        const leftChild: number = this.getLeftChild(sumId);
        const rightChild: number = this.getRightChild(sumId);

        const leftSum: number = this.getSum(start, middle, leftChild, queryStart, queryEnd);
        const rightSum: number = this.getSum(middle + 1, end, rightChild, queryStart, queryEnd);

        return leftSum + rightSum;
    }

    public updateIndex (id: number, val: number): null {
        if (id < 0 || id >= this.nums.length) {
            return null;
        }

        const diff: number = val - this.nums[id];
        this.nums[id] = val;

        this.update(0, this.nums.length - 1, 0, id, diff);

        return null;
    }

    public getSumOfRange (start: number, end: number): number {
        if (start < 0 || end >= this.nums.length || start > end) {
            return -1;
        }

        return this.getSum(0, this.nums.length - 1, 0, start, end);
    }
}

class NumArray {
    sTree: SegmentTree;

    constructor(nums: number[]) {
        this.sTree = new SegmentTree(nums);
    }

    update(index: number, val: number): void {
        this.sTree.updateIndex(index, val);
    }

    sumRange(left: number, right: number): number {
        return this.sTree.getSumOfRange(left, right);
    }
}



const test = () => {
    const params = [
        {
            input: [
                ["NumArray", "sumRange", "update", "sumRange"],
                [[1, 3, 5], [0, 2], [1, 2], [0, 2]]
            ],
            output: [null, 9, null, 8],
        },
        {
            input: [
                ["NumArray","sumRange","update","sumRange"],
                [[-1],[0,0],[0,1],[0,0]]
            ],
            output: [null, 9, null, 8],
        },
    ];

    params.forEach(({ input, output }) => {
        const arr: number[] = input[1][0] as number[];
        const na = new NumArray(arr);

        for (let i: number = 1; i < input[0].length; ++i) {
            const p: number[] = input[1][i] as number[];
            let result: number|null = null;

            if (input[0][i] === 'sumRange') {
                result = na.sumRange(p[0], p[1]);
            } else {
                na.update(p[0], p[1]);
            }

            console.log(
                result === output[i] ? 'SUCCESS ' : 'ERROR ',
                'input', JSON.stringify([input[0][i], input[1][i]]),
                'output', output[i],
                'result', result,
            );
        }
    });
};

test();

File: leetcode/medium/307_NumArray.py
from typing import List
import json
from collections import deque
import math


class SegemntTree:
    def __init__(self, nums):
        self.n = len(nums)
        length = 4 * self.n

        self.nums = nums
        self.array = [0] * length
        self.buildTree(0, 0, self.n - 1)

    def buildTree(self, id: int, start: int, end: int) -> int:
        if start == end:
            self.array[id] = self.nums[start]
            return self.array[id]

        lId = 1 + id * 2
        rId = 2 + id * 2
        middle = (start + end) // 2

        left = self.buildTree(lId, start, middle)
        right = self.buildTree(rId, middle + 1, end)
        self.array[id] = left + right
        return self.array[id]

    def _get(self, id, s, e, start, end):
        if s >= start and e <= end:
            return self.array[id]

        if e < start or s > end:
            return 0

        middle = (s + e) // 2
        lId = 1 + id * 2
        rId = 2 + id * 2

        left = self._get(lId, s, middle, start, end)
        right = self._get(rId, middle + 1, e, start, end)

        return left + right

    def get(self, start: int, end: int):
        s = 0
        e = self.n - 1
        return self._get(0, s, e, start, end)

    def _update(self, id, start, end, i, diff):
        if start > i:
            return
        if end < i:
            return

        self.array[id] += diff

        if start == end:
            return

        middle = (start + end) // 2
        lId = 1 + 2 * id
        rId = 2 + 2 * id

        self._update(lId, start, middle, i, diff)
        self._update(rId, middle + 1, end, i, diff)

        return

    def update(self, id: int, val: int):
        diff: int = val - self.nums[id]
        self.nums[id] = val

        return self._update(0, 0, self.n - 1, id, diff)


"""
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5], [6, 7, 8, 9]
[1, 2, 3], [4, 5], [6, 7], [8, 9]
[1, 2], [3], [4], [5], [6], [7], [8], [9]
[1], [2], [3], [4], [5], [6], [7], [8], [9]




"""


class NumArray:
    def __init__(self, nums: List[int]):
        self.segemntTree = SegemntTree(nums)

    def update(self, index: int, val: int) -> None:
        self.segemntTree.update(index, val)
        return

    def sumRange(self, left: int, right: int) -> int:
        return self.segemntTree.get(left, right)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)


def test():
    params = [
        # {
        #     "input": [
        #         ["NumArray", "sumRange", "update", "sumRange"],
        #         [[[1, 3, 5]], [0, 2], [1, 2], [0, 2]],
        #     ],
        #     "output": [None, 9, None, 8],
        # },
        {
            "input": [
                ["NumArray", "update", "sumRange", "sumRange", "update", "sumRange"],
                [[[9, -8]], [0, 3], [1, 1], [0, 1], [1, -3], [0, 1]],
            ],
            "output": [None, 9, None, 8, 1, 1],
        },
    ]

    for param in params:
        commands, data = param["input"]
        numArray = NumArray(*data[0])

        for i in range(1, len(commands)):
            result = None
            if commands[i] == "update":
                result = numArray.update(*data[i])
            if commands[i] == "sumRange":
                result = numArray.sumRange(*data[i])

            correct = json.dumps(result) == json.dumps(param["output"][i])

            msg = "SUCCESS" if correct else "ERROR"
            msg += "\n"
            msg += "input " + commands[i] + " " + json.dumps(data[i]) + "\n"
            msg += "output " + json.dumps(param["output"][i]) + "\n"
            msg += "result " + json.dumps(result) + "\n"

            print(msg)


if __name__ == "__main__":
    test()
